File: test/benchmark/larceny/divrec.js
// Larceny Benchmark: divrec (Node.js)
// Recursive integer division via repeated subtraction
'use strict';

function divrecDivHelper(x, y, q) {
    if (x < y) return q;
    return divrecDivHelper(x - y, y, q + 1);
}

function divrecDiv(x, y) {
    return divrecDivHelper(x, y, 0);
}

function divrecMod(x, y) {
    if (x < y) return x;
    return divrecMod(x - y, y);
}

function main() {
    const __t0 = performance.now();
    let result = 0;
    for (let iter = 0; iter < 1000; iter++) {
        result += divrecDiv(1000, 2);
        result -= divrecMod(1000, 2);
    }
    const __t1 = performance.now();
    if (result === 500000) {
        process.stdout.write("divrec: PASS\n");
    } else {
        process.stdout.write("divrec: FAIL result=" + result + "\n");
    }
    process.stdout.write("__TIMING__:" + (__t1 - __t0) + "\n");
}

main();

File: test/benchmark/larceny/python/divrec.py
#!/usr/bin/env python3
# Larceny Benchmark: divrec (Python)
# Recursive integer division via repeated subtraction
import sys
import time

sys.setrecursionlimit(100000)


def divrec_div(x, y):
    if x < y:
        return 0
    return 1 + divrec_div(x - y, y)


def divrec_mod(x, y):
    if x < y:
        return x
    return divrec_mod(x - y, y)


def main():
    t0 = time.perf_counter_ns()
    result = 0
    for _ in range(1000):
        result += divrec_div(1000, 2)
        result -= divrec_mod(1000, 2)
    t1 = time.perf_counter_ns()

    if result == 500000:
        print("divrec: PASS")
    else:
        print(f"divrec: FAIL result={result}")
    print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")


main()
